### PR TITLE
MODREQMED-23: Mediated request activities - add staff slips context

### PR DESCRIPTION
**Pull request by imerabishvili**
## Purpose
https://folio-org.atlassian.net/browse/MODREQMED-23

## Approach
Add clients needed too fetch corresponding objects from inventory. Use them to populate staffSlipContext for response.

#### TODOS and Open Questions
Need to populate **fromServicePoint**

### CODE REVIEW

File: src/main/java/org/folio/mr/service/impl/StaffSlipContextService.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    var materialType = inventoryService.fetchMaterialType(item.getMaterialTypeId());
+    var loanType = inventoryService.fetchLoanType(item.getPermanentLoanTypeId());
+    var toServicePoint = inventoryService.fetchServicePoint(item.getInTransitDestinationServicePointId());
+
+    var staffSlipContextItem = new SendItemInTransitResponseStaffSlipContextItem()
+      .title(instance.getTitle())
+      .primaryContributor(getPrimaryContributorName(instance))
+      .allContributors(getAllContributorNames(instance))
+
+      .barcode(item.getBarcode())
```
Comment (roman-barannyk) [inline src/main/java/org/folio/mr/service/impl/StaffSlipContextService.java:41]:
```suggestion
      .allContributors(getAllContributorNames(instance))
      .barcode(item.getBarcode())
```

File: src/main/java/org/folio/mr/service/impl/StaffSlipContextService.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+      .status(item.getStatus().getName().getValue())
+
+      .enumeration(item.getEnumeration())
```
Comment (roman-barannyk) [inline src/main/java/org/folio/mr/service/impl/StaffSlipContextService.java:44]:
```suggestion
      .status(item.getStatus().getName().getValue())
      .enumeration(item.getEnumeration())
```

File: src/main/java/org/folio/mr/service/impl/StaffSlipContextService.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+      .volume(item.getVolume())
+      .chronology(item.getChronology())
+      .copy(Objects.requireNonNullElse(item.getCopyNumber(), ""))
+      .displaySummary(item.getDisplaySummary())
+      .yearCaption(String.join("; ", item.getYearCaption()))
+
+      .materialType(materialType.getName())
```
Comment (roman-barannyk) [inline src/main/java/org/folio/mr/service/impl/StaffSlipContextService.java:51]:
```suggestion
      .yearCaption(String.join("; ", item.getYearCaption()))
      .materialType(materialType.getName())
```

File: src/main/java/org/folio/mr/service/impl/StaffSlipContextService.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+    if (item.getLastCheckIn() != null) {
+      staffSlipContextItem
+        .lastCheckedInDateTime(DateFormatUtil.formatUtcDate(item.getLastCheckIn().getDateTime()));
+    }
+
+    var location = inventoryService.fetchLocation(item.getEffectiveLocationId());
+    if (location != null) {
+
+      var library = inventoryService.fetchLibrary(location.getLibraryId());
```
Comment (roman-barannyk) [inline src/main/java/org/folio/mr/service/impl/StaffSlipContextService.java:67]:
```suggestion
    if (location != null) {
      var library = inventoryService.fetchLibrary(location.getLibraryId());
```

File: src/main/java/org/folio/mr/service/impl/StaffSlipContextService.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+        .effectiveLocationLibrary(library.getName())
+        .effectiveLocationCampus(campus.getName())
+        .effectiveLocationInstitution(institution.getName())
+        .effectiveLocationDiscoveryDisplayName(location.getDiscoveryDisplayName());
+
+      var primaryServicePoint = inventoryService.fetchServicePoint(location.getPrimaryServicePoint().toString());
+      if (primaryServicePoint != null) {
+        log.info("createStaffSlipContext:: primaryServicePoint is not null");
+        staffSlipContextItem
+          .effectiveLocationPrimaryServicePointName(primaryServicePoint.getName());
```
Comment (roman-barannyk) [inline src/main/java/org/folio/mr/service/impl/StaffSlipContextService.java:82]:
```suggestion
        staffSlipContextItem.effectiveLocationPrimaryServicePointName(
          primaryServicePoint.getName());
```

File: src/main/java/org/folio/mr/service/impl/StaffSlipContextService.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+      .copy(Objects.requireNonNullElse(item.getCopyNumber(), ""))
+      .displaySummary(item.getDisplaySummary())
+      .yearCaption(String.join("; ", item.getYearCaption()))
+
+      .materialType(materialType.getName())
+      .loanType(loanType.getName())
+      .numberOfPieces(item.getNumberOfPieces())
+      .descriptionOfPieces(item.getDescriptionOfPieces())
+
+      .fromServicePoint("TODO: missing service point?")
```
Comment (roman-barannyk) [inline src/main/java/org/folio/mr/service/impl/StaffSlipContextService.java:56]: lastCheckIn.servicePointId ?

File: src/main/java/org/folio/mr/service/impl/StaffSlipContextService.java (hunk 1 above)
Comment (imerabishvili) [reply src/main/java/org/folio/mr/service/impl/StaffSlipContextService.java]: done

File: src/main/java/org/folio/mr/service/impl/StaffSlipContextService.java (hunk 2 above)
Comment (imerabishvili) [reply src/main/java/org/folio/mr/service/impl/StaffSlipContextService.java]: done

File: src/main/java/org/folio/mr/service/impl/StaffSlipContextService.java (hunk 3 above)
Comment (imerabishvili) [reply src/main/java/org/folio/mr/service/impl/StaffSlipContextService.java]: done

File: src/main/java/org/folio/mr/service/impl/StaffSlipContextService.java (hunk 4 above)
Comment (imerabishvili) [reply src/main/java/org/folio/mr/service/impl/StaffSlipContextService.java]: done

File: src/main/java/org/folio/mr/service/impl/StaffSlipContextService.java (hunk 5 above)
Comment (imerabishvili) [reply src/main/java/org/folio/mr/service/impl/StaffSlipContextService.java]: done

File: src/main/java/org/folio/mr/service/impl/StaffSlipContextService.java (hunk 6 above)
Comment (imerabishvili) [reply src/main/java/org/folio/mr/service/impl/StaffSlipContextService.java]: Added - thank you!

**Review by roman-barannyk — APPROVED** [review]

File: src/main/java/org/folio/mr/service/impl/StaffSlipContextService.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+@RequiredArgsConstructor
+@Log4j2
+public class StaffSlipContextService {
+
+  private final InventoryService inventoryService;
+
+  public SendItemInTransitResponseStaffSlipContext createStaffSlipContext(MediatedRequest request) {
+    Item item = inventoryService.fetchItem(request.getItemId());
+
+    log.debug("createStaffSlipContext:: parameters item: {}", item);
```
Comment (alexanderkurash) [inline src/main/java/org/folio/mr/service/impl/StaffSlipContextService.java:27]: What if we were provided with invalid item ID? Let's make this log the first line of the method and log the request.

File: src/main/java/org/folio/mr/service/impl/StaffSlipContextService.java [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+    var holding = inventoryService.fetchHolding(item.getHoldingsRecordId());
+    var instance = inventoryService.fetchInstance(holding.getInstanceId());
+    var materialType = inventoryService.fetchMaterialType(item.getMaterialTypeId());
+    var loanType = inventoryService.fetchLoanType(item.getPermanentLoanTypeId());
+    var toServicePoint = inventoryService.fetchServicePoint(item.getInTransitDestinationServicePointId());
```
Comment (alexanderkurash) [inline src/main/java/org/folio/mr/service/impl/StaffSlipContextService.java:33]: Can any of these be null (including item)? If they can, could this cause an NPE?

Comment (imerabishvili) [reply src/main/java/org/folio/mr/service/impl/StaffSlipContextService.java:33]: What I found by looking into docs an tested with IT (MediatedRequestActionsApiTest) FeignException.NotFound will be thrown but not handled by ApiExceptionHandler. I will add FeignException handler:
image

File: src/main/java/org/folio/mr/service/impl/StaffSlipContextService.java (hunk 7 above)
Comment (imerabishvili) [reply src/main/java/org/folio/mr/service/impl/StaffSlipContextService.java:27]: done

**Review by alexanderkurash — APPROVED** [review]

**Pull request merged**